[path_provider_linux] Using TMPDIR env as a primary temporary path

proninyaroslav opened the pull request:
TMPDIR is a standard variable on UNIX/Linux systems, and is often used in containers such as Flatpak to redirect to a temporary folder inside a sandbox. This allows not to make hard bindings to the /tmp directory

Fixes https://github.com/flutter/flutter/issues/87742

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

stuartmorgan commented:
Thanks for the submission! This will need the remaining items in the checklist addressed before it moves forward with review. If you have any questions about completing those steps that aren’t addressed in the linked documentation, please let me know.

proninyaroslav commented:
@stuartmorgan 
I have questions about this: 
```
I updated pubspec.yaml with an appropriate new version according to the pub versioning philosophy.
I updated CHANGELOG.md to add a description of the change.
```
Do I have to make a version bump in this case?

proninyaroslav commented:
@stuartmorgan 
About tests: the problem is that the `Platform` class from the `dart:io` library is not mockable, so I can't insert the required environment variable value inside the test (`Platform.environment`). The solution would be to use the mockable version of `Platform`, some plugins use this https://pub.dev/packages/platform

stuartmorgan commented:
> I have questions about this:
> 
> ```
> I updated pubspec.yaml with an appropriate new version according to the pub versioning philosophy.
> I updated CHANGELOG.md to add a description of the change.
> ```
> 
> Do I have to make a version bump in this case?

The section you are quoting links to an explanation of the versioning policy for flutter/plugins. Could you elaborate on what your are seeing there that suggests you would not need to so that I can clarify it?

> About tests: the problem is that the `Platform` class from the `dart:io` library is not mockable, so I can't insert the required environment variable value inside the test (`Platform.environment`). The solution would be to use the mockable version of `Platform`, some plugins use this https://pub.dev/packages/platform

Since all you need in the `environment`, a much simpler solution at this point would just be to make `PathProviderLinux` take an injectable-by-tests environment map that would replace the call to `Platform.environment`. We generally try to minimize package dependencies in the plugins here.

proninyaroslav commented:
@stuartmorgan 
I added tests and also bumped the plugin version to 2.1.0 as I believe the API changes are backwards compatible and are not just a patch.

stuartmorgan commented on packages/path_provider/path_provider_linux/lib/path_provider_linux.dart:
```diff
@@ -12,14 +13,39 @@ import 'package:xdg_directories/xdg_directories.dart' as xdg;
+  /// Registers this class as the default instance of [PathProviderPlatform]
+  /// with the given [environment]
+  @visibleForTesting
+  static void privateRegisterWith({
```
Why does this method exist? `registerWith` is called automatically; nothing will call this automatically, so it's not clear what value this is providing over the tests just setting `instance` directly, and it seems actively confusing as it's named in the same style as something that *is* called automatically.

stuartmorgan commented on packages/path_provider/path_provider_linux/lib/path_provider_linux.dart:
```diff
@@ -12,14 +13,39 @@ import 'package:xdg_directories/xdg_directories.dart' as xdg;
   @override
   Future<String?> getTemporaryPath() {
-    return Future<String?>.value('/tmp');
+    final String? tmpdir = _environment['TMPDIR'];
```
How about `environmentTmpDir` as the name here, to make its role clearer. This current name doesn't communicate that it doesn't already include the fallback.

stuartmorgan commented on packages/path_provider/path_provider_linux/lib/path_provider_linux.dart:
```diff
@@ -12,14 +13,39 @@ import 'package:xdg_directories/xdg_directories.dart' as xdg;
-    return Future<String?>.value('/tmp');
+    final String? tmpdir = _environment['TMPDIR'];
+    return Future<String?>.value(
+      tmpdir == null || tmpdir.isEmpty ? '/tmp' : tmpdir,
```
You could simplify this slightly by making the variable above non-nullable with ` = _environment['TMPDIR'] ?? ''`, then just doing an `isEmpty` here.

stuartmorgan commented on packages/path_provider/path_provider_linux/test/path_provider_linux_test.dart:
```diff
@@ -7,7 +7,15 @@ import 'package:path_provider_platform_interface/path_provider_platform_interfac
 
 void main() {
   TestWidgetsFlutterBinding.ensureInitialized();
-  PathProviderLinux.registerWith();
```
You are removing critical test coverage here; this was ensuring that `registerWith` worked.

stuartmorgan commented on packages/path_provider/path_provider_linux/test/path_provider_linux_test.dart:
```diff
@@ -7,7 +7,15 @@ import 'package:path_provider_platform_interface/path_provider_platform_interfac
+    PathProviderLinux.privateRegisterWith(
+      environment: environment,
+    );
+  });
```
`setUp` should only be used when something is needed in multiple tests, which isn't the case here. You can just make a local `environment` and a local `PathProviderLinux` instance in a test specifically of environment variable handling.

stuartmorgan commented on packages/path_provider/path_provider_linux/test/path_provider_linux_test.dart:
```diff
@@ -16,6 +24,12 @@ void main() {
   test('getTemporaryPath', () async {
     final PathProviderPlatform plugin = PathProviderPlatform.instance;
     expect(await plugin.getTemporaryPath(), '/tmp');
```
Unless I'm missing something, this will fail if someone runs it on a machine with TMPDIR set to something non-default. If you want to explicitly test null, you should be using an environment created by the test that doesn't contain TMPDIR, rather than assuming the host environment doesn't have it.

proninyaroslav replied:
We pass `Map<String, String>` as environment , which is initially empty. Therefore, there can be no conflict.

proninyaroslav commented:
@stuartmorgan 
All comments have been resolved.

proninyaroslav commented:
@stuartmorgan 
Can you review my changes again?

stuartmorgan reviewed (approved):
Sorry for the delay. Looking again, the tests would be much clearer as three distinct tests rather than one test that alternates between changing the environment and then making new assertions. Rather than do another round of review, I just made those change in the PR.

LGTM with that. Thanks!

stuartmorgan commented:
Oops, this got caught by the Windows bot change so it was never auto-landed. Landing manually, since this can't affect Windows.
